stable_baselines3/common/recurrent/policies.py

```diff
@@ -209,42 +221,73 @@ def _process_sequence(
         lstm_output = th.flatten(th.cat(lstm_output).transpose(0, 1), start_dim=0, end_dim=1)
         return lstm_output, lstm_states
 
-    def forward(
+    def recurrent_initial_state(self, n_envs: Optional[int] = None, *, device: Optional[th.device | str] = None):
+        shape: tuple[int, ...]
+        if n_envs is None:
+            shape = (self.lstm_hidden_state_shape[0], self.lstm_hidden_state_shape[2])
+        else:
+            shape = (self.lstm_hidden_state_shape[0], n_envs, self.lstm_hidden_state_shape[2])
+        return RNNStates(
+            (th.zeros(shape, device=device), th.zeros(shape, device=device)),
+            (th.zeros(shape, device=device), th.zeros(shape, device=device)),
+        )
+
+    # Methods for getting `latent_vf` or `latent_pi`
+    def _recurrent_latent_pi_and_vf(
+        self, obs: TorchGymObs, state: RNNStates, episode_starts: th.Tensor
+    ) -> Tuple[Tuple[th.Tensor, th.Tensor], RNNStates]:
+        features = self.extract_features(obs)
+        pi_features: th.Tensor
+        vf_features: th.Tensor
+        if self.share_features_extractor:
+            assert isinstance(features, th.Tensor)
+            pi_features = vf_features = features
+        else:
+            assert isinstance(features, tuple)
+            pi_features, vf_features = features
+        latent_pi, lstm_states_pi = self._process_sequence(pi_features, state.pi, episode_starts, self.lstm_actor)
+        latent_vf, lstm_states_vf = self._recurrent_latent_vf_from_features(vf_features, state, episode_starts)
+        if lstm_states_vf is None:
+            lstm_states_vf = (lstm_states_pi[0].detach(), lstm_states_pi[1].detach())
+        return ((latent_pi, latent_vf), RNNStates(lstm_states_pi, lstm_states_vf))
+
+    def _recurrent_latent_vf_from_features(
+        self, vf_features: th.Tensor, state: RNNStates, episode_starts: th.Tensor
+    ) -> Tuple[th.Tensor, Optional[LSTMStates]]:
+        "Get only the vf features, not advancing the hidden state"
+        if self.lstm_critic is None:
+            if self.shared_lstm:
+                with th.no_grad():
```

Comment:
I would have thought that we need this `with th.no_grad():` at the top level (ie, at line 257 and applying to all parts of this function).

In particular, I'm wondering if we might accidentally alter gradients on line 266 otherwise.